components/movementsensor/gpsnmea/serial.go

```diff
@@ -180,7 +183,10 @@ func (g *SerialNMEAMovementSensor) Accuracy(ctx context.Context, extra map[strin
 func (g *SerialNMEAMovementSensor) LinearVelocity(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
 	g.mu.RLock()
 	defer g.mu.RUnlock()
-	return r3.Vector{X: 0, Y: g.data.Speed, Z: 0}, nil
+	headingInRadians := g.data.CompassHeading * (math.Pi / 180)
+	xVelocity := g.data.Speed * math.Sin(headingInRadians)
+	yVelocity := g.data.Speed * math.Cos(headingInRadians)
+	return r3.Vector{X: xVelocity, Y: yVelocity, Z: 0}, g.err.Get()
```

Comment:
This code all looks copied and pasted from the other file. Can it be refactored into a shared helper function that gets called from two different places? That way, we only need to maintain one version of the code: you only need to fix bugs once, you only need to add new features once, and there's no chance that the two implementations diverge and future engineers need to wonder if they're supposed to be the same.

Same suggestion about the compass heading stuff.

Reply:
Susmita and I discussed: this code can be moved into the `GPSData` struct instead, so that this function becomes:
```go
func {
    g.mu.RLock()
    defer g.mu.RUnlock()
    return g.data.CalculateTheThing(), g.err.Get()
}
```
However, Susmita would prefer to do all the cleanup stuff in a separate PR. This one is for adding a bunch of spare code, and that future one is for cleaning it all up into something better. I guess that's alright with me: it seems like a more complicated approach, but if this is what Susmita prefers, the end result is something we both agree on.